optuna/integration/wandb.py

```diff
@@ -0,0 +1,84 @@
+from typing import Any
+from typing import Dict
+from typing import Optional
+from typing import Union
+
+import optuna
+from optuna._experimental import experimental
+from optuna._imports import try_import
+
+
+with try_import() as _imports:
+    import wandb
+
+
+@experimental("2.9.0")
+class WeightsAndBiasesCallback(object):
+    """Callback to track Optuna trials with Weights & Biases.
+
+    This callback enables tracking of Optuna study in
+    Weights & Biases. The study is tracked as a single experiment
+    run, where all suggested hyperparameters and optimized metric
+    are logged and plotted as a function of optimizer steps.
+
+    .. note::
+        User needs to be logged in to Weights & Biases before
+        using this callback in online mode. For more information, please
+        refer to `wandb setup <https://docs.wandb.ai/quickstart#1-set-up-wandb>`_.
+
+    Example:
+
+        Add Weights & Biases callback to Optuna optimization.
+
+        .. code::
+
+            import optuna
+            from optuna.integration.wandb import WeightsAndBiasesCallback
+
+
+            def objective(trial):
+                x = trial.suggest_float("x", -10, 10)
+                return (x - 2) ** 2
+
+
+            wandb_kwargs = {"project": "my-project"}
+            wandbc = WeightsAndBiasesCallback(wandb_kwargs=wandb_kwargs)
+
+
+            study = optuna.create_study(study_name="my_study")
+            study.optimize(objective, n_trials=10, callbacks=[wandbc])
+
+    Args:
+        metric_name:
+            Name of the optimized metric. Since the metric itself is just a number,
+            ``metric_name`` can be used to give it a name. So you know later
+            if it was roc-auc or accuracy.
+        wandb_kwargs:
+            Set of arguments passed when initializing Weights & Biases run.
+            Please refer to `Weights & Biases API documentation
+            <https://docs.wandb.ai/ref/python/init>`_ for more details.
+    """
+
+    def __init__(
+        self, metric_name: Union[str] = "value", wandb_kwargs: Optional[Dict[str, Any]] = None
+    ) -> None:
+
+        _imports.check()
+
+        self._metric_name = metric_name
+        self._wandb_kwargs = wandb_kwargs or {}
+
+        self._initialize_run()
+
+    def __call__(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -> None:
+
+        direction = str(study.direction).split(".")[-1]
+        attributes = {"direction": direction}
+
+        wandb.config.update(attributes)
+        wandb.log({**trial.params, self._metric_name: trial.value}, step=trial.number)
```

Comment:
`trial.number` must only increase to be able to use the `step=trial.number`.  If it's possible for `trial.number` to decrease I would suggest something like:

```
commit = self._last_trial_number != trial.number
wandb.log({**trial.params, self._metric_name: trial.value, "global_step": trial.number}, commit=commit)
self._last_trial_number = trial.number
```

Reply:
The addition of the `commit` logic is only needed if it's common for this function to be called many times for the same `trial.number`.  If not, you can just get rid of the commit logic.

Reply:
@nzw0301 please correct me if I'm wrong, but assuming sequential optimization, this should not be a problem, right?

Reply:
Hi, @vanpelt, thank you for giving your feedback! In this PR, we can assume that `trial.number` must increase. However, in the future, we support such a situation when `trial.number` might decrease, especially, distributed optimisation by optuna. So your feedback is really helpful!

Reply:
Ah, so it would be great to mention it as a note as well. Optuna still supports the `n_jobs` argument of `study.optimize` function. Even though we use a single process optimization, it might be happen when `n_jobs != 1`.

Reply:
I will add a note on this as well. Thanks @vanpelt for the feedback!